solr/core/src/java/org/apache/solr/cli/StreamTool.java

```diff
@@ -0,0 +1,507 @@
+          StringBuilder outLine = new StringBuilder();
+          if (outputHeaders != null) {
+            for (int i = 0; i < outputHeaders.length; i++) {
+              if (i > 0) {
+                outLine.append(delimiter);
+              }
+
+              Object o = tuple.get(outputHeaders[i]);
+              if (o != null) {
+                if (o instanceof List) {
+                  List outfields = (List) o;
+                  outLine.append(listToString(outfields, arrayDelimiter));
+                } else {
+                  outLine.append(o);
+                }
+              }
+            }
+          }
+          CLIO.out(outLine.toString());
+        }
+      }
+    } finally {
+
+      if (pushBackStream != null) {
+        pushBackStream.close();
+      }
+
+      solrClientCache.close();
+    }
+
+    echoIfVerbose("StreamTool -- Done.");
+  }
+
+  public PushBackStream doLocalMode(CommandLine cli, String expr) throws Exception {
+    String zkHost = SolrCLI.getZkHost(cli);
+
+    echoIfVerbose("Connecting to ZooKeeper at " + zkHost);
+    solrClientCache.getCloudSolrClient(zkHost);
+    solrClientCache.setBasicAuthCredentials(cli.getOptionValue(CommonCLIOptions.CREDENTIALS_OPTION));
+
+    TupleStream stream;
+    PushBackStream pushBackStream;
+
+    StreamExpression streamExpression = StreamExpressionParser.parse(expr);
+    StreamFactory streamFactory = new StreamFactory();
+
+    // stdin is ONLY available in the local mode, not in the remote mode as it
+    // requires access to System.in
+    streamFactory.withFunctionName("stdin", StandardInStream.class);
+
+    // LocalCatStream extends CatStream and disables the Solr cluster specific
+    // logic about where to read data from.
+    streamFactory.withFunctionName("cat", LocalCatStream.class);
+
+    streamFactory.withDefaultZkHost(zkHost);
+
+    Lang.register(streamFactory);
+
+    stream = StreamTool.constructStream(streamFactory, streamExpression);
+
+    pushBackStream = new PushBackStream(stream);
+
+    // Now we can run the stream and return the results.
+    StreamContext streamContext = new StreamContext();
+    streamContext.setSolrClientCache(solrClientCache);
+
+    // Output the headers
+    pushBackStream.setStreamContext(streamContext);
+
+    return pushBackStream;
+  }
+
+  public PushBackStream doRemoteMode(CommandLine cli, String expr) throws Exception {
+
+    String solrUrl = SolrCLI.normalizeSolrUrl(cli);
+    if (!cli.hasOption("name")) {
+      throw new IllegalStateException(
+          "You must provide --name COLLECTION with --worker solr parameter.");
+    }
+    String collection = cli.getOptionValue("name");
+
+    if (expr.toLowerCase(Locale.ROOT).contains("stdin(")) {
+      throw new IllegalStateException(
+          "The stdin() expression is only usable with --worker local set up.");
+    }
+
+    final SolrStream solrStream =
+        new SolrStream(solrUrl + "/solr/" + collection, params("qt", "/stream", "expr", expr));
```

Comment:
Don't like string concats for URLs, because we get into "v1 / v2" dilemma, but I am not sure if there is another way for this?